### PR TITLE
EZP-30753: Changed loadSections method to filter available sections

### PR DESCRIPTION
**Pull request by mikadamczyk**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-30753](https://jira.ez.no/browse/EZP-30753)
| **Bug/Improvement**| yes
| **New feature**    |no
| **Target version** | `7.5`
| **BC breaks**      | yes
| **Tests pass**     | yes/no
| **Doc needed**     | yes/no

Changed `loadSection()` method to filter the available sections and not throw an exception when the user has no permission to all sections. This change is needed to allow display user a list of section available to him in the administration panel.

**TODO**:
- [x] Implement feature / fix a bug.
- [x] Implement tests.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

File: eZ/Publish/API/Repository/SectionService.php [hunk 1]
```diff
@@ -57,9 +57,7 @@ public function updateSection(Section $section, SectionUpdateStruct $sectionUpda
     public function loadSection($sectionId);
 
     /**
-     * Loads all sections.
-     *
-     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException If the current user user is not allowed to read a section
+     * Loads all sections. If the current user is not allowed to read a section then list are filtered.
```
Comment (alongosz) [inline eZ/Publish/API/Repository/SectionService.php]: Maybe something like:
```suggestion
     * Loads all sections, excluding the ones the current user is not allowed to read.
```

File: eZ/Publish/API/Repository/Tests/SectionServiceAuthorizationTest.php [hunk 2]
```diff
@@ -120,10 +122,9 @@ public function testUpdateSectionThrowsUnauthorizedException()
      * Test for the loadSections() method.
      *
      * @see \eZ\Publish\API\Repository\SectionService::loadSections()
-     * @expectedException \eZ\Publish\API\Repository\Exceptions\UnauthorizedException
      * @depends eZ\Publish\API\Repository\Tests\SectionServiceTest::testLoadSections
      */
-    public function testLoadSectionsThrowsUnauthorizedException()
+    public function testLoadSectionDidNotThrowAnException()
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Tests/SectionServiceAuthorizationTest.php]:
```suggestion
    public function testLoadSectionsLoadsEmptyListForAnonymousUser()
```

File: eZ/Publish/API/Repository/Tests/SectionServiceAuthorizationTest.php [hunk 3]
```diff
@@ -149,9 +150,57 @@ public function testLoadSectionsThrowsUnauthorizedException()
         // Set anonymous user
         $repository->setCurrentUser($userService->loadUser($anonymousUserId));
 
-        // This call will fail with a "UnauthorizedException"
-        $sectionService->loadSections();
+        $sections = $sectionService->loadSections();
+        /* END: Use Case */
+
+        // Sections should be filtered to those which user has access
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Tests/SectionServiceAuthorizationTest.php]:
```suggestion
```
I'd skip this, with the above change code is self-commenting enough.

File: eZ/Publish/API/Repository/Tests/SectionServiceAuthorizationTest.php [hunk 4]
```diff
@@ -149,9 +150,57 @@ public function testLoadSectionsThrowsUnauthorizedException()
...
+            'MediaUser',
+            new Limitation\SectionLimitation(['limitationValues' => [$expectedSection->id]])
+        );
+
+        $repository->getPermissionResolver()->setCurrentUserReference($mediaUser);
+
+        $sections = $sectionService->loadSections();
         /* END: Use Case */
+
+        // Sections should be filtered to those which user has access
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Tests/SectionServiceAuthorizationTest.php]:
```suggestion
        // Only Sections the user has access to should be loaded
```

File: eZ/Publish/Core/REST/Client/SectionService.php [hunk 5]
```diff
@@ -156,9 +156,7 @@ public function loadSection($sectionId)
     }
 
     /**
-     * Loads all sections.
-     *
-     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException If the current user user is not allowed to read a section
+     * Loads all sections. If the current user is not allowed to read a section then list are filtered.
```
Comment (alongosz) [inline eZ/Publish/Core/REST/Client/SectionService.php]:
```suggestion
     * Loads all sections, excluding the ones the current user is not allowed to read.
```

File: eZ/Publish/Core/Repository/SectionService.php [hunk 6]
```diff
@@ -203,25 +204,19 @@ public function loadSection($sectionId)
     }
 
     /**
-     * Loads all sections.
-     *
-     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException If the current user user is not allowed to read a section
+     * Loads all sections. If the current user is not allowed to read a section then list are filtered.
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/SectionService.php]:
```suggestion
     * Loads all sections, excluding the ones the current user is not allowed to read.
```

File: eZ/Publish/Core/SignalSlot/SectionService.php [hunk 7]
```diff
@@ -119,9 +119,7 @@ public function loadSection($sectionId)
     }
 
     /**
-     * Loads all sections.
-     *
-     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException If the current user user is not allowed to read a section
+     * Loads all sections. If the current user is not allowed to read a section then list are filtered.
```
Comment (alongosz) [inline eZ/Publish/Core/SignalSlot/SectionService.php]:
```suggestion
     * Loads all sections, excluding the ones the current user is not allowed to read.
```

File: eZ/Publish/Core/Repository/SectionService.php [hunk 8]
```diff
@@ -203,25 +204,19 @@ public function loadSection($sectionId)
...
-        $sections = [];
-        foreach ($this->sectionHandler->loadAll() as $spiSection) {
-            $sections[] = $section = $this->buildDomainSectionObject($spiSection);
-
-            // @todo change API to just filter instead of throwing here
-            if (!$this->permissionResolver->canUser('section', 'view', $section)) {
-                throw new UnauthorizedException('section', 'view');
-            }
-        }
+        $sections = array_map(function ($spiSection) {
```
Comment (webhdx) [inline eZ/Publish/Core/Repository/SectionService.php]: `array_map([$this, 'buildDomainSectionObject'], $this->sectionHandler->loadAll());`

**Comment by lserwatka** [comment]
You can merge it up

**Comment by micszo** [comment]
Tickets (draft state): https://jira.ez.no/browse/EZP-30783, https://jira.ez.no/browse/EZP-30784.

**Pull request merged**